crates/algebra/src/state_machine.rs

```diff
@@ -0,0 +1,180 @@
+    /// the StateMachine.
+    pub fn state_hook<R, Fut: Sync + Send + Future<Output = ()> + 'static>(
+        &mut self,
+        mut f: impl FnMut(&State) -> Fut + 'static,
+    ) {
+        self.state_hook = Box::new(move |s: &State| f(s).boxed());
+    }
+}
+
+impl<
+        Input: 'static,
+        Config: Unpin,
+        State: Default + Unpin,
+        Output: Unpin,
+        Err,
+        F: FnMut(&Config, &mut State, Input) -> Result<Output, Err> + Unpin,
+    > Stream for StateMachine<Input, Config, State, Output, Err, F>
+{
+    type Item = Result<Output, Err>;
+
+    fn poll_next(
+        self: std::pin::Pin<&mut Self>,
+        cx: &mut std::task::Context<'_>,
+    ) -> std::task::Poll<Option<Self::Item>> {
+        let StateMachine {
+            input_stream,
+            config,
+            state,
+            stf,
+            state_hook,
+            pending_state_hook,
+            ..
+        } = std::pin::Pin::into_inner(self);
+
+        // First check if we still have to poll the State continuation.
+        if let Some((mut fut, out)) = pending_state_hook.take() {
+            return match fut.as_mut().poll(cx) {
+                std::task::Poll::Ready(()) => std::task::Poll::Ready(Some(Ok(out))),
+                std::task::Poll::Pending => {
+                    pending_state_hook.replace((fut, out));
+                    std::task::Poll::Pending
+                }
+            };
+        }
+
+        // Otherwise we can pull another event off of the upstream and process it.
+        match Pin::new(input_stream).poll_next(cx) {
+            std::task::Poll::Ready(None) => std::task::Poll::Ready(None),
+            std::task::Poll::Ready(Some(input)) => match stf(config, state, input) {
+                Ok(output) => {
+                    let mut state_hook_fut = state_hook(state);
+                    match state_hook_fut.as_mut().poll(cx) {
+                        std::task::Poll::Ready(()) => std::task::Poll::Ready(Some(Ok(output))),
+                        std::task::Poll::Pending => {
+                            *pending_state_hook = Some((state_hook_fut, output));
+                            std::task::Poll::Pending
+                        }
+                    }
+                }
+                Err(err) => std::task::Poll::Ready(Some(Err(err))),
+            },
+            std::task::Poll::Pending => std::task::Poll::Pending,
+        }
+    }
+}
+
+#[cfg(test)]
+mod state_machine_tests {
+    use futures::stream::StreamExt;
+
+    use super::*;
+    #[tokio::test]
+    async fn test_sm() {
+        let mut sm = StateMachine::restore((), 0, test_stf);
+
+        let inputs = stream::StreamExt::cycle(stream::iter(vec![false, true])).take(10);
+
+        sm.attach(inputs);
+
+        let transcript = sm
+            .filter_map(|x| future::ready(x.ok()))
+            .collect::<Vec<usize>>()
+            .await;
+        println!("{transcript:?}");
```

Comment:
Use logging and not `println` statements in tests.
Also, can you add an `assert!`/`assert_eq` here?